### PR TITLE
BUG: EAs should not be hashable

### PR DESCRIPTION
**Pull request by jbrockmendel**
This turns out to be a blocker for some cleanup in index.pyx

### CODE REVIEW

**Comment by gfyoung** [comment]
LGTM - do we need a `whatsnew` ?

**Comment by WillAyd** [comment]
Do we actually want to commit to this? Just wonder if this would conflict with some of the work I think @TomAugspurger did before to try and create a non-mutable dataframe

**Comment by jreback** [comment]
yes mutable things are not hashable
we do allow hash_pandas_object to work though

**Comment by TomAugspurger** [comment]
This change seems fine. Can you say more about the motivation though? As I can imagine an immutable array that is hasable, so we mange shouldn’t rely on unhashabilty.

**Comment by jbrockmendel** [comment]
> Can you say more about the motivation though? As I can imagine an immutable array that is hasable, so we mange shouldn’t rely on unhashabilty.

I'm working on indexing code and want to simplify a check

```
cdef inline bint is_definitely_invalid_key(object val):
    if isinstance(val, tuple):
        try:
            hash(val)
        except TypeError:
            return True

    # we have a _data, means we are a NDFrame
    return (isinstance(val, slice) or util.is_array(val)
            or isinstance(val, list) or hasattr(val, '_data'))

    cpdef get_loc(self, object val):
        cdef:
            Py_ssize_t loc

        if is_definitely_invalid_key(val):
            raise TypeError(f"'{val}' is an invalid key")
    [...]
```

to just check for hashability.  Doing that breaks a couple of tests because EAs are currently hashable

**Comment by TomAugspurger** [comment]
> to just check for hashability. Doing that breaks a couple of tests because EAs are currently hashable

When an immutable, 3rd-party EA comes along that's hashable, will that break again?

**Comment by jbrockmendel** [comment]
> When an immutable, 3rd-party EA comes along that's hashable, will that break again?

If that EA gets passed as a `key` to `index._engine.(get_loc|get_value|contains)` then it may give `(KeyError|KeyError|False)` instead of raising a `TypeError`

File: pandas/core/arrays/base.py [hunk 1]
```diff
@@ -1073,6 +1073,9 @@ def _reduce(self, name, skipna=True, **kwargs):
         """
         raise TypeError(f"cannot perform {name} with type {self.dtype}")
 
+    def __hash__(self):
```
Comment (jreback) [inline pandas/core/arrays/base.py]: may want to add some text to the top of the EA that says EA by default are not hashable

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**